src/runtime/auto-prompt-manager.js

```diff
@@ -0,0 +1,172 @@
+/**
+ * Copyright 2021 The Subscribe with Google Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {AutoPromptType} from '../api/basic-subscriptions';
+import {assert} from '../utils/log';
+
+/**
+ * Manages the display of subscription/contribution prompts automatically
+ * displayed to the user.
+ */
+export class AutoPromptManager {
+  /**
+   * @param {!./deps.DepsDef} deps
+   */
+  constructor(deps) {
+    /** @private @const {!./deps.DepsDef} */
+    this.deps_ = deps;
+
+    /** @private @const {!../model/page-config.PageConfig} */
+    this.pageConfig_ = deps.pageConfig();
+
+    /** @private @const {!./entitlements-manager.EntitlementsManager} */
+    this.entitlementsManager_ = deps.entitlementsManager();
+
+    /** @private @const {?./client-config-manager.ClientConfigManager} */
+    this.clientConfigManager_ = deps.clientConfigManager();
+    assert(
+      this.clientConfigManager_,
+      'AutoPromptManager requires an instance of ClientConfigManager.'
+    );
+  }
+
+  /**
+   * Triggers the display of the auto prompt, if preconditions are met.
+   * Preconditions are as follows:
+   *   - alwaysShow == true, used for demo purposes, OR
+   *   - There is no active entitlement found AND
+   *   - The user had not reached the maximum impressions allowed, as specified
+   *     by the publisher
+   * A prompt may not be displayed if the appropriate criteria are not met.
+   * @param {{
+   *   autoPromptType: (string|undefined),
+   *   alwaysShow: (boolean|undefined),
+   * }=} options
+   * @param {function()=} displayForLockedContentFn
```

Comment:
Could the `displayForLockedContentFn` param be part of the `options` object?

Reply:
It could, options were meant to contain the mini prompt configuration details, passed in from the BasicSubscriptions API, and displayForLockedContentFn kind of fell outside of that because it provides the alternative-to-mini-prompt display route. I don't feel too strongly about this and could see combining them to a more general params object though, so let me know if you'd like to see them combined. Just out of curiosity, is the way the parameters are currently structured against a design pattern?

Reply:
I'd love to see them combined, I just think that would simplify things

Reply:
Done.